benchmarks/send-recv-core.py

```diff
@@ -0,0 +1,541 @@
+        op_completed()
+
+    def _tag_recv_handle(request, exception, ep, msg):
+        assert exception is None
+        req = ucx_api.tag_send_nb(
+            ep, msg, msg.nbytes, tag=0, cb_func=_send_handle, cb_args=(msg,)
+        )
+        if req is None:
+            op_completed()
+
+    def _am_recv_handle(recv_obj, exception, ep):
+        assert exception is None
+        msg = Array(recv_obj)
+        ucx_api.am_send_nbx(ep, msg, msg.nbytes, cb_func=_send_handle, cb_args=(msg,))
+
+    def _listener_handler(conn_request, msg):
+        global ep
+        ep = ucx_api.UCXEndpoint.create_from_conn_request(
+            worker,
+            conn_request,
+            endpoint_error_handling=True,
+        )
+
+        # Wireup before starting to transfer data
+        if args.enable_am is True:
+            ucx_api.am_recv_nb(ep, cb_func=_am_recv_handle, cb_args=(ep,))
+        else:
+            wireup = Array(bytearray(len(WireupMessage)))
+            op_started()
+            ucx_api.tag_recv_nb(
+                worker,
+                wireup,
+                wireup.nbytes,
+                tag=0,
+                cb_func=_tag_recv_handle,
+                cb_args=(ep, wireup),
+            )
+
+        for i in range(args.n_iter):
+            if args.enable_am is True:
+                ucx_api.am_recv_nb(ep, cb_func=_am_recv_handle, cb_args=(ep,))
+            else:
+                if not args.reuse_alloc:
+                    msg = Array(xp.zeros(args.n_bytes, dtype="u1"))
+
+                op_started()
+                ucx_api.tag_recv_nb(
+                    worker,
+                    msg,
+                    msg.nbytes,
+                    tag=0,
+                    cb_func=_tag_recv_handle,
+                    cb_args=(ep, msg),
+                )
+
+    if not args.enable_am and args.reuse_alloc:
+        msg = Array(xp.zeros(args.n_bytes, dtype="u1"))
+    else:
+        msg = None
+
+    listener = ucx_api.UCXListener(
+        worker=worker, port=args.port or 0, cb_func=_listener_handler, cb_args=(msg,)
+    )
+    queue.put(listener.port)
+
+    while outstanding[0] == 0:
+        worker.progress()
+
+    # +1 to account for wireup message
+    if args.delay_progress:
+        while finished[0] < args.n_iter + 1 and (
+            outstanding[0] >= args.max_outstanding
+            or finished[0] + args.max_outstanding >= args.n_iter + 1
+        ):
+            worker.progress()
+    else:
+        while finished[0] != args.n_iter + 1:
+            worker.progress()
+
+
+def client(queue, port, server_address, args):
+    if args.client_cpu_affinity >= 0:
+        os.sched_setaffinity(0, [args.client_cpu_affinity])
+
+    import numpy as np
+
+    if args.object_type == "numpy":
+        import numpy as xp
+    elif args.object_type == "cupy":
+        import cupy as xp
+
+        xp.cuda.runtime.setDevice(args.client_dev)
+    else:
+        import cupy as xp
+
+        import rmm
```

Comment:
Could the imports go to the top of the file?

Reply:
Sorry, this file is here by accident, it's the old file that was previously removed after the benchmarks were refactored, but I accidentally reintroduced it after merging the new branch due to my original changes to this file in the PR. This has been removed again in https://github.com/rapidsai/ucx-py/pull/836/commits/0fae0b696f33fe66217cf3d40735de166849a0e0 .